content/posts/2024-11-05-flang-new.md

```diff
@@ -0,0 +1,963 @@
+
+> What did I hope to see as a result of the name change? I wanted it to be
+> easier for more people to try it out.
+
+So once you have finished reading this article, download Flang or try it out
+on Compiler Explorer. You know at least one person will appreciate it!
+
+TODO: download link goes here
+TODO: CE link goes here.
+
+# Fortran Intermediate Representation (FIR)
+
+All compilers that use LLVM as a backend eventually produce code in the form of
+the [LLVM Intermediate Representation](https://llvm.org/docs/LangRef.html)
+(LLVM IR).
+
+A drawback of this is that LLVM IR does not include language specific information
+. This means that for example, it cannot be used to optimise arrays in a way
+specific to Fortran programs.
+
+One solution to this has been to build a higher level IR that represents the
+unique features of the language, optimise that, then convert the result into LLVM IR.
+
+Eric Schweitz started to do that for Fortran in late 2018:
+
+> FIR was originally conceived as a high-level IR that would interoperate with
+> LLVM but have a representation more friendly and amenable to Fortran
+> optimizations.
+
+Naming is hard but Eric defied expectations:
+
+> FIR was a pun of sorts. Fortran IR and meant to be evocative of the trees
+> (Abstract Syntax Trees).
+
+We will not go into detail about this early FIR, because [MLIR](https://mlir.llvm.org/)
+was revealed soon after Eric started the project and they quickly adopted it.
+
+> When MLIR was announced, I quickly switched gears from building data
+> structures for a new "intermediate IR" to porting my IR design to MLIR and
+> using that instead.
+>
+> I believe FIR was probably the first "serious project" outside of Google to
+> start using MLIR.
+
+The FIR work continued to develop, with Jean Perier (Nvidia) joining Eric on
+the project. It became its own public branch `fir-dev`, that was latercontributed to the LLVM Project.
+
+The following sections will go into detail on the IRs that Flang uses today.
+
+# MLIR
+
+The journey from Classic Flang to LLVM Flang involved a rewrite of the
+entire compiler. This provided an opportunity to pick up new things from
+the LLVM Project. Most notably MLIR.
+
+"Multi-Level Intermediate Representation" (MLIR) was first
+[introduced](https://www.youtube.com/watch?v=qzljG6DKgic) to the LLVM
+community in 2019 (around the time that F18 was approved to move into the LLVM Project).
+
+The problem that MLIR addresses is the same one that Eric Schweitz tackled with FIR.
+It is difficult to map high level details of programming languages
+into LLVM IR.
+
+You either have to to attach them to the IR as metadata, try to recover the
+lost details later, or fight an uphill battle to add the details to
+LLVM IR itself. These details are crucial for producing optimised code in certain
+languages (Fortran array optimisations were an early cited use case for MLIR).
+
+This lead languages such as Swift and Rust to create their own IRs that include
```

Comment:
```suggestion
This led languages such as Swift and Rust to create their own IRs that include
```

Reply:
Fixed, also found another instance of lead -> led, it's one of my blind spots clearly.